### PR TITLE
Merge multiple settlements in driver

### PR DESCRIPTION
**Pull request by vkgnosis**
Finally enabled merging of multiple settlements.

How do you think we should test this? I could adapt the e2e test so that we assert that in a single run it settles 4 orders instead of the current two.

### CODE REVIEW

File: solver/src/driver.rs [hunk 1]
```diff
@@ -314,14 +337,24 @@ struct RatedSettlement {
...
+        .next()
+        .ok_or_else(|| anyhow!("no settlements"))?;
+    for settlement in settlements.take(MAX_MERGED_SETTLEMENTS - 1) {
+        merged = merged.merge(settlement)?;
+    }
+    Ok(merged)
+}
+
 #[cfg(test)]
 mod tests {
```
Comment (nlordell) [inline solver/src/driver.rs:352]: Might be worth adding some unit tests for the new logic.

**Review by nlordell — COMMENTED** [review]
Changes look good.

I would like to see a couple of unit tests that demonstrates how merging affects the `SolverWIthSettlements` result, but it may be quite verbose to setup. Maybe if more of the code in the `filter_map` closure were in a function, it might be easier to test?

File: solver/src/driver.rs [hunk 2]
```diff
@@ -314,14 +337,24 @@ struct RatedSettlement {
     objective_value: BigRational,
 }
 
+// Goes through the settlements in order and tries to merge at most MAX_SETTLEMENT_MERGES together.
+fn merge_settlements(mut settlements: impl Iterator<Item = Settlement>) -> Result<Settlement> {
+    let mut merged = settlements
+        .next()
+        .ok_or_else(|| anyhow!("no settlements"))?;
+    for settlement in settlements.take(MAX_MERGED_SETTLEMENTS - 1) {
+        merged = merged.merge(settlement)?;
```
Comment (fleupold) [inline solver/src/driver.rs:346]: If a specific settlement cannot be merged into the existing bunch, shouldn't we still try to merge the rest of the settlements? This way we can still reduce the number of total transactions we are waiting for.

File: solver/src/driver.rs [hunk 3]
```diff
@@ -23,6 +23,8 @@ use std::{
 // protect against a gas estimator giving bogus results that would drain all our funds.
 const GAS_PRICE_CAP: f64 = 500e9;
 
+const MAX_MERGED_SETTLEMENTS: usize = 5;
```
Comment (fleupold) [inline solver/src/driver.rs:26]: Should this be configurable via command line so it's easier to adapt?

**Review by fleupold — APPROVED** [review]
Second the suggestion to add unit tests. The driver logic is becoming more and more complex and doesn't yet have good coverage.

Otherwise, great PR 🎉 !

**Review by bh2smith — APPROVED** [review]

File: solver/src/driver.rs (hunk 2 above)
Comment (vkgnosis) [reply solver/src/driver.rs]: Fixed.

**Comment by vkgnosis** [comment]
I added two more commits which:
* move some of the settlement handling logic into another file
* split out merge related function and added unit test
* make max merge count configurable

File: solver/src/driver.rs [hunk 4]
```diff
@@ -204,25 +175,31 @@ impl Driver {
...
-                    settlement.settlement,
-                );
-            }
-        }
+            .run_solvers(liquidity, gas_price)
+            .await
+            .filter_map(solver_settlements::filter_bad_settlements)
+            .inspect(|(name, settlements)| {
+                for settlement in settlements.iter() {
+                    tracing::debug!("solver {} found solution:\n {:?}", name, settlement,);
```
Comment (nlordell) [inline solver/src/driver.rs:183]: ubernit: trailing comma
```suggestion
                    tracing::debug!("solver {} found solution:\n {:?}", name, settlement);
```

File: solver/src/driver/solver_settlements.rs [hunk 5]
```diff
@@ -0,0 +1,204 @@
+use crate::settlement::Settlement;
+use anyhow::Result;
+use num::BigRational;
+use primitive_types::H160;
+use std::{collections::HashMap, time::Duration};
+
+// Return None if the result is an error or there are no settlements remaining after removing
+// settlements with no trades.
+pub fn filter_bad_settlements(
```
Comment (nlordell) [inline solver/src/driver/solver_settlements.rs:9]: "bad" - so judgemental :stuck_out_tongue:.

**Review by nlordell — APPROVED** [review]
Looks amazing

**Pull request merged**